rego/rego.go

```diff
@@ -45,14 +46,62 @@ type PartialResult struct {
 	compiler *ast.Compiler
 	store    storage.Store
 	body     ast.Body
+	r        *Rego
 }
 
 // Rego returns an object that can be evaluated to produce a query result.
+// If rego.Rego#Prepare was used to create the PartialResult this may lose
+// the pre-parsed/compiled parts of the original Rego object. In those cases
+// using rego.PartialResult#Eval is likely to be more performant.
 func (pr PartialResult) Rego(options ...func(*Rego)) *Rego {
 	options = append(options, Compiler(pr.compiler), Store(pr.store), ParsedQuery(pr.body))
 	return New(options...)
 }
 
+// EvalOptions defines the set of options allowed to be set at evaluation
+// time. Any other options will need to be set on a new Rego object.
+type EvalOptions struct {
+	Input       interface{}
+	Metrics     metrics.Metrics
+	Transaction storage.Transaction
+}
+
+// Eval evaluates this PartialResult's Rego object with additional eval options
+// and returns a ResultSet.
+// If options are provided they will override the original Rego options respective value.
+func (pr PartialResult) Eval(ctx context.Context, options *EvalOptions) (ResultSet, error) {
```

Comment:
Are we allowing options to be nil in case the caller doesn't need to provide any arguments? Just trying to understand why we're passing by reference instead of value here. This is just a question and doesn't have to affect the PR.

Reply:
Yep, that was pretty much it. Makes the "just eval it" call go from `pr.Eval(ctx, rego.EvalOptions{})` to `pr.Eval(ctx, nil)` so nothing more than a convenience. I could go either way on it, no strong preference.